src/Pate/Verification/PairGraph.hs

```diff
@@ -0,0 +1,878 @@
+
+       oSeq <- liftIO (MT.observableEvents sym oMem)
+       pSeq <- liftIO (MT.observableEvents sym pMem)
+
+       traceBundle bundle $ unlines
+         [ "== original event trace =="
+         , show (prettySymSequence ppEvent oSeq)
+         ] 
+
+       traceBundle bundle $ unlines
+         [ "== patched event trace =="
+         , show (prettySymSequence ppEvent pSeq)
+         ]
+
+       -- TODO! actually check the equivalance of the observables
+
+
+-- TODO move into MemTrace and do it properly
+ppEvent :: LCB.IsSymInterface sym => MT.MemEvent sym ptrW -> Doc ann
+ppEvent (MT.MemOpEvent op)   = pretty "MemOp"
+ppEvent (MT.SyscallEvent ex) = pretty "SyscallEvent" <+> W4.printSymExpr ex
+
+
+data TotalityResult
+  = CasesTotal
+  | TotalityCheckingError String
+  | TotalityCounterexample (Integer,MS.MacawBlockEndCase) (Integer,MS.MacawBlockEndCase)
+
+-- TODO? should we try to share work with the followExit/widenPostcondition calls?
+checkTotality :: forall sym arch.
+  W4.Pred sym ->
+  SimBundle sym arch ->
+  AbstractDomain sym arch ->
+  [PPa.PatchPair (PB.BlockTarget arch)] ->
+  EquivM sym arch TotalityResult
+checkTotality asm bundle preD exits =
+  withSym $ \sym ->
+    do vcfg <- asks envConfig
+       eqRel <- asks envBaseEquiv
+       stackRegion <- asks (PMC.stackRegion . envCtx)
+
+       let solver = PCfg.cfgSolver vcfg
+       let saveInteraction = PCfg.cfgSolverInteractionFile vcfg
+
+       precond <- liftIO $ do
+         eqInputs <- PE.getPredomain sym stackRegion bundle eqRel (PS.specBody preD)
+         W4.andPred sym asm eqInputs
+
+       -- compute the condition that leads to each of the computed
+       -- exit pairs
+       cases <- forM exits $ \(PPa.PatchPair oBlkt pBlkt) ->
+                  PD.matchesBlockTarget bundle oBlkt pBlkt
+
+       isUnknown <- do
+         isJump <- matchingExits bundle MS.MacawBlockEndJump
+         isFail <- matchingExits bundle MS.MacawBlockEndFail
+         isBranch <- matchingExits bundle MS.MacawBlockEndBranch
+         liftIO (W4.orPred sym isJump =<< W4.orPred sym isFail isBranch)
+
+       isReturn <- do
+         bothReturn <- matchingExits bundle MS.MacawBlockEndReturn
+         abortO <- PAb.isAbortedStatePred (PPa.getPair @PBi.Original (simOut bundle))
+         returnP <- liftIO $ MS.isBlockEndCase (Proxy @arch) sym (PS.simOutBlockEnd $ PS.simOutP bundle) MS.MacawBlockEndReturn
+         abortCase <- liftIO $ W4.andPred sym abortO returnP
+         liftIO $ W4.orPred sym bothReturn abortCase
+
+
+       let doPanic = panic Solver "checkTotality" ["Online solving not enabled"]
+
+       PS.withOnlineSolver solver saveInteraction sym $ \bak ->
+           do liftIO $ LCBO.withSolverProcess bak doPanic $ \sp -> do
+                W4.assume (W4.solverConn sp) precond
+                W4.assume (W4.solverConn sp) =<< W4.notPred sym isReturn
+                W4.assume (W4.solverConn sp) =<< W4.notPred sym isUnknown
```

Comment:
Does this need to be in a new solver frame?

Reply:
`withOnlineSolver` starts a fresh solver instance .  If we were using a more persistent solver, then yes, I think a solver frame here would be necessary.